Implement untagged unions (RFC 1444)

petrochenkov opened the pull request:
cc https://github.com/rust-lang/rust/issues/32836

Notes:
- The RFC doesn't talk about `#[packed]` unions, this implementation supports them, packing changes union's alignment to 1 and removes trailing padding.
- The RFC doesn't talk about dynamically sized unions, this implementation doesn't support them and rejects them during wf-checking (similarly, dynamically sized enums are not supported as well).
- The lint for drop fields in unions can't work precisely before monomorphization, so it works pessimistically - non-`Copy` generic fields are reported, types not implementing `Drop` directly, but having non-trivial drop code are reported.

    ```
    struct S(String); // Doesn't implement `Drop`
    union U<T> {
        a: S, // Reported
        b: T, // Reported
    }
    ```

- https://github.com/rust-lang/rust/pull/35764 was indeed helpful and landed timely, I didn't have to implement internal drop flags for unions.
- Unions are not permitted in constant patterns, because matching on union fields is unsafe, I didn't want unsafety checker to dig into all constants to uncover this possible unsafety.
- The RFC doesn't talk about `#[derive]`, generally trait impls cannot be derived for unions, but some of them can. I implemented only `#[derive(Copy)]` so far. In theory shallow `#[derive(Clone)]` can be derived as well if all union fields are `Copy`, I left it for later though, it requires changing how `Clone` impls are generated.
- Moving union fields is implemented as per https://github.com/rust-lang/rust/issues/32836#issuecomment-242511491.
- Testing strategy: union specific behavior is tested, sometimes very basically (e.g. debuginfo), behavior common for all ADTs (e.g. something like coherence
checks) is not generally tested.

r? @eddyb

petrochenkov commented:
Actually, I'd like to make the implementation a bit more conservative, than written in the RFC.
Namely, to replace the `unions_with_drop_fields` lint with a hard error and require user-defined `Drop` implementation for unions with potentially non-trivially destructible fields.
Effectively, `#[allow(unions_with_drop_fields)]` is replaced with `impl Drop for U { fn drop(&mut self) {} }` in code wanting to use such unions.

eddyb commented on src/librustc_mir/build/expr/as_rvalue.rs:
```diff
@@ -181,6 +181,9 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
             } => { // see (*) above
+                let is_union = adt_def.adt_kind() == ty::AdtKind::Union;
+                let active_field_index = if is_union { Some(fields[0].name.index()) } else { None };
+
```
It might be a better idea to use direct assignment here instead of an `Rvalue` - that would remove the need to change `AggregateKind::Adt` - alternatively, an `Union` `AggregateKind` could be added.

petrochenkov replied:
>alternatively, an Union AggregateKind could be added

IIRC, I tried this and didn't like the result (too much common boilerplate code for `Adt` and `Union`)

eddyb replied:
The field assignment approach might be best then.

petrochenkov replied:
Not sure what it means. Any example?

eddyb replied:
Instead of `lvalue = union { f: operand }` you'd have `lvalue.f = operand`.

eddyb replied:
Just realized that could make `{ let u: Union; u.f = x; u }` pass a future MIR move checker - do we want that?

petrochenkov replied:
Future move checker with full support for fragments should accept both `{ let u: Union; u.f = x; u }` and `{ let s: StructWith1Field; s.f = x; s }` I guess, see the last item in https://github.com/rust-lang/rust/pull/36016#issuecomment-242867725.

eddyb replied:
Oh, then the field approach is definitely correct - except for custom `Drop` impls, which require whole-value initialization "to be armed", what's the plan there?

nikomatsakis replied:
> Oh, then the field approach is definitely correct - except for custom Drop impls, which require whole-value initialization "to be armed", what's the plan there?

There is special treatment for types that implement drop, yes. The idea is that you can't have such a type in a "partially initialized" state, so you can't move out from its fields, nor can you write to a field if the type itself is not initialized.

eddyb commented:
LGTM, great job!
I'm a bit sad about adding `TyUnion` and supporting `trans::adt` but I think it's easier to land this PR as-is and punt on those long-standing duplication issues.

r? @nikomatsakis for the borrowck bits & miscellaneous other details.

petrochenkov commented:
>I think it's easier to land this PR as-is and punt on those long-standing duplication issues.

That would be nice, this patch set bitrots relatively quickly.
Improvements/refactorings in trans can be made together with refactoring remains of the old trans (also, I don't have enough expertize in the area to do polishing).

nikomatsakis commented on src/librustc/middle/effect.rs:
```diff
@@ -178,11 +177,28 @@ impl<'a, 'tcx, 'v> Visitor<'v> for EffectCheckVisitor<'a, 'tcx> {
                 }
             }
+            hir::ExprField(ref base_expr, field) => {
+                if let ty::TyUnion(..) = self.tcx.expr_ty_adjusted(base_expr).sty {
```
Nit: I feel like it'd be nice to `match` the sty (exhaustively) and have a `bug!` call for the unexpected cases. This makes me mildly nervous that some bug or addition might slip through.

nikomatsakis commented on src/librustc/middle/effect.rs:
```diff
@@ -178,11 +177,28 @@ impl<'a, 'tcx, 'v> Visitor<'v> for EffectCheckVisitor<'a, 'tcx> {
+
+    fn visit_pat(&mut self, pat: &hir::Pat) {
+        if let PatKind::Struct(_, ref fields, _) = pat.node {
+            if let ty::TyUnion(..) = self.tcx.pat_ty(pat).sty {
```
same here; I wonder if we should factor this into some form of helper where you just pass in the type?

petrochenkov replied:
This case and the case above can't be unified reasonably, pattern fields and expr fields are different types.

nikomatsakis commented on src/librustc/middle/expr_use_visitor.rs:
```diff
@@ -670,30 +669,33 @@ impl<'a, 'gcx, 'tcx> ExprUseVisitor<'a, 'gcx, 'tcx> {
-                        with_field.ty(self.tcx(), substs)
-                    );
-                    self.delegate_consume(with_expr.id, with_expr.span, cmt_field);
+        match with_cmt.ty.sty {
```
Is there any change here, or it's just converting `if let` to `match`?

petrochenkov replied:
>Is there any change here

Not in the resulting diff, I'll change this back to `if let`.

nikomatsakis commented on src/librustc/mir/repr.rs:
```diff
@@ -962,7 +962,10 @@ pub enum CastKind {
+    /// The second field is variant number (discriminant), it's equal to 0
+    /// for struct and union expressions. The fourth field is active field
+    /// number and is present only for union expressions.
+    Adt(AdtDef<'tcx>, usize, &'tcx Substs<'tcx>, Option<usize>),
```
Hmm. Interesting. I sort of expected the union "field" to translate into an ADT variant, though I can see why that would make things inconvenient.

petrochenkov replied:
Yeah, it's a consequence of unions being univariant `AdtDef`s, the second `usize` is used for indexing into `AdtDef`'s variants all the time, it's more convenient to keep it `0`, like for `struct`s.
(I'm not proud of this, but this is the less invasive change.)

eddyb replied:
Also see https://github.com/rust-lang/rust/pull/36016/files#r76464770.

nikomatsakis commented on src/librustc/ty/mod.rs:
```diff
@@ -1433,6 +1435,7 @@ bitflags! {
         const IS_PHANTOM_DATA     = 1 << 3,
         const IS_SIMD             = 1 << 4,
         const IS_FUNDAMENTAL      = 1 << 5,
+        const IS_UNION            = 1 << 7,
```
any particular reason not to use `1 << 6`?

petrochenkov replied:
Oops, bad rebase, `1 << 6` was `IS_NO_DROP_FLAG`.

nikomatsakis commented on src/librustc_trans/debuginfo/metadata.rs:
```diff
@@ -1203,6 +1216,66 @@ fn prepare_tuple_metadata<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
+// Unions
+//=-----------------------------------------------------------------------------
+
+struct UnionMemberDescriptionFactory<'tcx> {
```
ping @michaelwoerister -- does this file look roughly right? :)

michaelwoerister replied:
Looks pretty good to me!

nikomatsakis commented on src/librustc_trans/glue.rs:
```diff
@@ -267,7 +267,8 @@ pub fn implement_drop_glue<'a, 'tcx>(ccx: &CrateContext<'a, 'tcx>,
 
 fn trans_struct_drop<'blk, 'tcx>(bcx: Block<'blk, 'tcx>,
```
can we rename this to `trans_adt_drop`?

eddyb replied:
IMO it should be `trans_custom_dtor`.

petrochenkov replied:
`trans_custom_dtor` fits better, adts with "inherited" drop code are processed elsewhere.

nikomatsakis commented on src/libsyntax_ext/deriving/clone.rs:
```diff
@@ -80,6 +80,7 @@ pub fn expand_deriving_clone(cx: &mut ExtCtxt,
         additional_bounds: bounds,
         generics: LifetimeBounds::empty(),
         is_unsafe: false,
+        supports_unions: false,
```
ideally, we would support `#[derive(Copy, Clone)]` but not `#[derive(Clone)]` in isolation

petrochenkov replied:
There are some difficulties with the current implementation of `#[derive(Clone)]`, it generates pattern matching even for the "shallow clone" case to check bounds, but these patterns are not appropriate for unions, this needs to be done in some other way somehow in some other PR. cc @durka 
```
#[derive(Copy, Clone)]
struct S {
	a: u8,
	b: u8,
}
```
=>
```
impl ::std::clone::Clone for S {
    #[inline]
    fn clone(&self) -> S {
        match *self {

            S { a: ref __self_0_0, b: ref __self_0_1 } => { // <== BAD

                ::std::clone::assert_receiver_is_clone(&(*__self_0_0));
                ::std::clone::assert_receiver_is_clone(&(*__self_0_1));
                *self
            }
        }
    }
}
impl ::std::marker::Copy for S { }
struct S {
    a: u8,
    b: u8,
}
```

nikomatsakis commented on src/test/compile-fail/union-borrow-nested.rs:
```diff
@@ -0,0 +1,44 @@
+
+// ignore-tidy-linelength
+
+#![feature(untagged_unions)]
```
can we rename this test to `borrowck-union-nested` (and/or put in the `borrowck` directory, if I made one of those already, I can't rememeber).

Also, can we put a coment saying what pattern is being tested here.

nikomatsakis replied:
Also, let's make this a `ui` test. Ideally, make a new directory `ui/borrowck`. See `src/test/ui/README.md` for advice on generating the ref files.

nikomatsakis commented on src/test/compile-fail/union-borrow.rs:
```diff
@@ -0,0 +1,97 @@
+
+// ignore-tidy-linelength
+
+#![feature(untagged_unions)]
```
Same deal w/ making a `ui` test and changing name to `borrowck/union-borrow.rs`

nikomatsakis commented on src/test/compile-fail/union-empty.rs:
```diff
@@ -0,0 +1,15 @@
+
+#![feature(untagged_unions)]
+
+union U {} //~ ERROR unions cannot have zero fields
```
Is there a particular reason for this limitation? Seems like something we'll inevitably lift later. Ah well, it's ok for now.

petrochenkov replied:
>Is there a particular reason for this limitation?

The RFC says empty unions are not allowed :)

>Seems like something we'll inevitably lift later.

Most probably.

joshtriplett replied:
I didn't have a strong reason for disallowing this, and I have no objection to dropping that limitation.  I don't remember what motivated me to explicitly mention that in the RFC; I seem to recall having a reason, but I don't remember the reason.

petrochenkov replied:
The semantics is unclear, is `union U {}` an uninhabited type like `enum E {}` or an empty type like `struct E {}`.
Both can be implemented by either keeping the "union expressions should have exactly one field" rule or relaxing it to "union expressions should have at most one field".

nikomatsakis commented on src/test/compile-fail/union-field-privacy.rs:
```diff
@@ -0,0 +1,21 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
```
For all these tests with names like `union-foo`, let's make a `union/` directory. (Also, I think we should just go ahead and move them all to `ui` tests, per the logic in [this thread](https://internals.rust-lang.org/t/testing-the-compiler/3836).)

nikomatsakis replied:
well, we can wait on the `ui` stuff -- we can do it all at once, probably. But let's definitely make a directory for `union` tests. (Though in some cases, e.g., with borrowck, and maybe things like privacy, I'd prefer to group `union` in with that.)

nikomatsakis commented on src/test/run-pass/union-backcomp.rs:
```diff
@@ -0,0 +1,23 @@
+    union U {
+        a: u8,
+    }
+}
```
nice.

nikomatsakis commented:
OK, did a review. Here are my thoughts:

- I was skeptical at first, but adding the `Option<usize>` to the MIR ADT does seem to work well. And I remember now that I was thinking of union as more of a `repr` attribute on struct, and that approach seems to fit with that interpretation (i.e., a struct where all fields are at offset 0). So I am happy with it.
- The borrowck integration is not the way I imagined doing it. I had imagined that instead of adding a loan for every field whenever one is borrowed, we would instead change the equality metric so that if you accessed `field.x` and `field.y` is borrowed where `x` and `y` potentially overlap, we would error out. 
    - That said, this way is not bad.
    - But I am unhappy about the dummy types and changing the def'n of `Eq` and so forth. I always think that a custom `Eq` is a sign of subtle logic and this seems to be no exception. Now you have to track which `LoanPath` instances have "real" types and which don't.
    - *But* I do plan to be rewriting this to operate over MIR and maybe we can address it then.
    - Have to think about this just a bit.
- The test cases are looking pretty good. As has been my practice for big PRs like this, I am going to go through the RFC and try to make a list of important points in there. I'll concatenate that with a list of tests I wrote down while skimming through the code and post it to make sure we hit all the important points. Many of them you already covered in any case.
- I agree with @eddyb that it'd be better to have `TyAdt` but also that this would be better done in a follow-up PR.

petrochenkov commented:
@nikomatsakis 
>And I remember now that I was thinking of union as more of a repr attribute on struct

This is exactly how it turned out to be from implementation point of view. Unions tend to stay close to their syntactic surface representation and therefore treated in the same way as structs by most of compiler passes except for layout calculation and borrow/move checking.
I can imagine if union's syntax were modeled after enums, all the implementation would tend to enums instead.

nikomatsakis commented:
OK here is my master list of tests. Feel free to check-off ones that you feel you already have (I have already done so in some cases, but I know you've got a lot of these covered already).

- [ ] test defining and calling an inherent method on a union type
- [ ] test suggestions when accessing a union field but different field with similar name was intended (e.g., user did `x.principle` instead of `x.principal`).
- [ ] test suggestions when accessing a union field but method was intended (e.g., user wrote `x.calculate` instead of `x.calculate()`
- [ ] test semantics of assigning directly to a field whose type implements drop
- [ ] test semantics of assigning to an enum which implements drop
    - enum dtor runs, fields should not be dropped
- [ ] test dropping an enum which has a custom dtor by doing out of scope
    - fields should not be dropped
- [ ] test dropping an enum which does not have a custom dtor by doing out of scope
    - fields should not be dropped
- [ ] test pattern matching using 1 field, as intended
- [ ] test pattern matching using 0 fields is error
- [ ] test pattern matching using >1 field is error
- [ ] test pattern matching using `..` is error
- [x] test union containing unsized types, both in tail and not in tail position
- [ ] test `repr(C)` applied to union and then passed to C code
- [ ] test that "c type" lint warns if union does *not* have `repr(C)`
- [ ] test that you can implement `Copy` even if all fields do not implement `Copy`
- test borrowck interactions
    - [ ] borrow `a`, write to `b`
    - [ ] borrow `a` mutably, borrow `b` shared
    - [ ] borrow `a` mutably, borrow `b` mutably
    - [x] move from `a` etc 
- [x] test rustdoc integration
- [ ] test that writing field is unsafe
- [ ] test that reading field is unsafe
- [ ] test that pattern matching union is unsafe
- [ ] test that constructing a union is not unsafe (though honestly...well, wev)
- [ ] test that union fields can be declared `pub` and then used outside of declaring module
- [ ] test that private union fields cannot be used outside of declaring module
    - extra credit: test `pub(restricted)` :)
- [ ] test that feature gate is required to declare a union type
- test generic unions:
    - [ ] run-pass with generic type parameters being used in various ways
    - [ ] run-pass test generic union w/ field containing a projection (`union Foo<T: Iterator> { elem: T::Item }`)
    - [ ] compile-fail with a where-clause that is not supposed (e.g., `union Foo<T: Copy>`, check that `Foo<Rc<u32>>` is an error to use)
- [x] test unions cannot be used in constant patterns (this is not in RFC, but seems good)
- [x] packed union size/alignment tests
- [ ] test instantiating a union with 0 or >1 fields yields error
- [ ] test lint warning if union has a `Drop` field
- [ ] test implementing some non-special trait for a union type and calling methods on it
- [x] test size/alignment is as expected
- [ ] test we can implement `Copy` on unions that have fields that are not `Copy`
    - we may want to lint on this too though? Seems good.
- [ ] test C interop (i.e., declare a `#[repr(C)]` union with a non-trivial layout and pass it to some C code)

Here are some tests we could add, except that I think the impl will not work quite the way I would want (and I'm ok with that):

- [ ] test that `derive(Copy, Clone)` works (or open a FIXME and leave a note on the tracking issue)
- [ ] test deferred union initialization
    - the rfc says `let x: Union; x.f = 22; use(x);` should work; however, this doesn't work right for structs afaik, so I think it's ok if it doesn't for now

nikomatsakis commented:
One thing I thought of while doing this: we probably want to consider the borrowck error messages. i.e., I suspect they could give better advice (in general we try to point out when a restriction on path `x` is due to a borrow of some other path `y`). But we can leave this for later.

petrochenkov commented:
>The borrowck integration is not the way I imagined doing it.

What I wrote is just the simplest way to implement it :D
But also it's the simplest way to explain/document the behavoir - "if one union's field is borrowed/moved/assigned then all its other fields are borrowed/moved/assigned as well".

>But I am unhappy about the dummy types and changing the def'n of Eq and so forth.

That assert in `PartialEq` had to be removed anyway (see https://github.com/rust-lang/rust/pull/36004), types can be slightly different even without unions.

nikomatsakis commented:
Having thought about the borrowck -- I think I'm inclined to land it with this approach and consider patching it up and refactoring it later. This PR is too good to stall for long. :)

nikomatsakis commented:
r=me once tests are set.

petrochenkov commented:
>One thing I thought of while doing this: we probably want to consider the borrowck error messages. i.e., I suspect they could give better advice (in general we try to point out when a restriction on path x is due to a borrow of some other path y). 

This is already done (kind of) - all additional loans for sibling union fields have the correct source loan path set (see `compile-fail/union-borrow.rs`), I'm not sure why not all error messages use this information.
```
let ra = &u.a;
let rmb = &mut u.b; //~ ERROR cannot borrow `u` (via `u.b`) as mutable because `u` is also borrowed as immutable (via `u.a`)
```

joshtriplett commented:
@petrochenkov Regarding the lint for unions with drop fields, note that several people specifically wanted to use a union to implement a `ManuallyDrop` wrapper.  So if you want to change that, you should check with some of those developers and see how it would impact them.

eddyb commented:
@joshtrplett It's a lint so you can allow it just for that `ManuallyDrop`.

joshtriplett commented:
@eddyb Agreed, but there was a mention of turning that into an error.  Personally, I don't have a problem with making the lint error-by-default rather than warn-by-default, but I think it still makes sense to allow overriding it without having to provide a no-op Drop impl for the union.  But I suggested talking to people who want `ManuallyDrop` about any possible changes there.

eddyb commented:
@joshtriplett You can still allow it if it's error-by-defsult

joshtriplett commented:
@eddyb I wasn't sure that "keep the lint and make it error-by-default" was what was being suggested, as opposed to "get rid of the lint and make it always an error unless you implement Drop for the union".

joshtriplett commented:
@nikomatsakis A couple of tests in your checklist say "enum" where they should say "union".

A few more tests:

- [ ] Test that disabling the lint for unions with drop fields works (to make sure the warning actually depends on the lint).
- [ ] Test pattern-matching a union as a field of a struct.
- [ ] Test that simultaneous borrows of multiple fields of a struct contained within a union do not conflict (test case included in RFC 1444).
- [ ] Test putting a union inside a struct marked as `packed`.  (For instance, define a packed struct containing a u16 and u8, a union containing that struct and a u16, and a packed struct containing a u8 followed by that union or vice versa=; the outermost struct should have size 4.)
- [ ] Test passing a union declaration to a macro.
- [ ] Test the contextual nature of the `union` keyword: declare and call a function named `union`, declare and use a variable named `union`.

bors commented:
:umbrella: The latest upstream changes (presumably #36030) made this pull request unmergeable. Please resolve the merge conflicts.

petrochenkov commented:
I've addressed all comments and added tests from both @nikomatsakis' and @joshtriplett's lists.

What is not implemented:
- Translate union rvalues into field assignments (https://github.com/rust-lang/rust/pull/36016#discussion_r76464770). Haven't investigated this yet.
- Tests are not converted into UI form (following this comment https://github.com/rust-lang/rust/pull/36016#discussion_r76495268).
- Regarding

    >test that you can implement Copy even if all fields do not implement Copy

    I would prefer not to implement unchecked `Copy`, at least not until people start actively requesting it.
Unchecked `Copy` breaks guarantees I described in https://github.com/rust-lang/rust/issues/32836#issuecomment-242814894 and closes the door to safer unions. If unchecked `Copy` is implemented, then all move checker support for unions can be thrown away as well, basically.
    (I've added a test checking the opposite behavior.)
- I haven't implemented my suggestion from https://github.com/rust-lang/rust/pull/36016#issuecomment-242810887, I'll probably submit an amendment to the RFC first.

@nikomatsakis, @eddyb, please confirm you agree with landing the PR in its current form.

joshtriplett commented:
One additional test that occurs to me after re-reading the RFC:
- [ ] Assigning to one field of a union does not overwrite data outside that field.

Quoting RFC 1444:
> If a union contains multiple fields of different sizes, assigning to a field smaller than the entire union must not change the memory of the union outside that field.

So, how about a union of a u8 and a struct containing a pair of u8 fields, assigning to both u8 fields of the struct, assigning to the u8 field of the union, and checking the other field of the struct?  (Repeat for u16, u32, and u64.)

This will make sure that rustc and LLVM don't turn a union field write into a larger memory write.

(Note that rustc/LLVM can turn an aggregate assignment of a union into a larger write, just not a single union field write.)

Thanks for all your work on this!

eddyb commented:
@petrochenkov LGTM.

petrochenkov commented:
@joshtriplett 
>So, how about a union of a u8 and a struct containing a pair of u8 fields, assigning to both u8 fields of the struct, assigning to the u8 field of the union, and checking the other field of the struct?

Added, `run-pass/union/union-overwrite.rs`

joshtriplett commented:
@petrochenkov Thanks!

LGTM.

Manishearth commented:
Is this ready to merge?

petrochenkov commented:
@Manishearth 
>Is this ready to merge?

Yes, it was `r=me`'d by @nikomatsakis  https://github.com/rust-lang/rust/pull/36016#issuecomment-242868051, and I added tests.
The unclear detail is `Copy` impls (https://github.com/rust-lang/rust/pull/36016#issuecomment-242948929), but the implemented behavior is forward compatible with anything.

petrochenkov commented:
Problems found by buildbot:
- [ ] Something is wrong with cross crate usage of unions (https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/2304/steps/test/logs/stdio)
- [ ] Debuginfo test fails with lldb (https://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/2278/steps/test/logs/stdio)

petrochenkov commented:
Problems fixed, waiting for https://github.com/rust-lang/rust/pull/36066 to land first.

joshtriplett commented:
@petrochenkov 
Responding to a couple of the items in the initial message in this thread:

> The RFC doesn't talk about #[packed] unions, this implementation supports them, packing changes union's alignment to 1 and removes trailing padding.

That's exactly the correct interpretation.

> The RFC doesn't talk about dynamically sized unions, this implementation doesn't support them and rejects them during wf-checking (similarly, dynamically sized enums are not supported as well).

That seems completely reasonable.  Including a dynamically sized type in a union seems non-trivial and would require quite a bit of additional specification not included in the RFC.  Prohibiting that seems fine.  If someone wants it, they should file an additional RFC with a detailed specification.

> The lint for drop fields in unions can't work precisely before monomorphization, so it works pessimistically - non-Copy generic fields are reported, types not implementing Drop directly, but having non-trivial drop code are reported.

The latter seems fine: types with non-trivial drop code should produce the same warning.

The former seems unfortunate.  It probably won't cause too much of a problem in practice, but I would have hoped that the warning would occur at the time of instantiating the union with a specific type that implements Drop, rather than at the time of declaring the union.  Nonetheless, if that adds too much complexity to rustc, in practice I don't see too much of a problem with requiring a `Copy` bound on generic types used in a union.

> Unions are not permitted in constant patterns, because matching on union fields is unsafe, I didn't want unsafety checker to dig into all constants to uncover this possible unsafety.

Can you elaborate on what this would mean in practice?  Does this mean I can't write the following?

```rust
union U {
    i: i32,
    f: f32,
}

struct S {
    d: u32,
    u: U,
}

fn f(s: S) {
    unsafe {
        match s {
            S { d: 0, u: U { i: 42 } } => { ... }
        }
    }
}
```

Ideally, I could write the above, and rustc could compile it into a single 64-bit comparison.

> The RFC doesn't talk about #[derive], generally trait impls cannot be derived for unions, but some of them can. I implemented only #[derive(Copy)] so far. In theory shallow #[derive(Clone)] can be derived as well if all union fields are Copy, I left it for later though, it requires changing how Clone impls are generated.

I don't see any sensible way to support deriving traits other than Copy and Clone for a union.  (Compiler extensions, and user-defined derive implementations could potentially work, though.)

petrochenkov commented:
@joshtriplett 
>>Unions are not permitted in constant patterns, because matching on union fields is unsafe, I didn't want unsafety checker to dig into all constants to uncover this possible unsafety.

>Can you elaborate on what this would mean in practice? Does this mean I can't write the following?

Things like this are prohibited:
```
union U {
    a: u8,
    b: u8,
}

const C: U = U { a: 10 }; // <-- here we hide union expression in a constant

fn main() {
    match C {
        C => {} // <-- here we are trying to match on that constant
        _ => {}
    }
}
```

joshtriplett commented:
@petrochenkov Ah, I see.  That seems reasonable; the "special case" of allowing structural equality for a type that doesn't implement PartialEq shouldn't apply to unions.

bors commented:
:umbrella: The latest upstream changes (presumably #36066) made this pull request unmergeable. Please resolve the merge conflicts.

petrochenkov commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit ae358c9 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis ae358c9dedbb1a522d106760ccad0c3161b9f9a4 -->

bors commented:
:umbrella: The latest upstream changes (presumably #36126) made this pull request unmergeable. Please resolve the merge conflicts.

petrochenkov commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit 7914856 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 7914856bf56ddd55d76ad2bab55c284a9f97d502 -->

bors commented:
:umbrella: The latest upstream changes (presumably #34623) made this pull request unmergeable. Please resolve the merge conflicts.

petrochenkov commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit 5dd423a has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 5dd423aaf3b9033b75bf5e76e6a76b345582d142 -->

bors commented:
:umbrella: The latest upstream changes (presumably #34494) made this pull request unmergeable. Please resolve the merge conflicts.

petrochenkov commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit 2f6cc28 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 2f6cc28849966e4affdd5c4ba8f53260aab9d990 -->

bors commented:
:umbrella: The latest upstream changes (presumably #35894) made this pull request unmergeable. Please resolve the merge conflicts.

petrochenkov commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit 79d9491 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 79d94916e5893cb8f19f0e7627fa923709195630 -->

bors commented:
:hourglass: Testing commit 79d9491 with merge 10e2631...

bors commented:
:broken_heart: Test failed - [auto-mac-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/2342)

petrochenkov commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit c424190 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis c42419048c96554e7778e62064b7a11d1945699d -->

bors commented:
:hourglass: Testing commit c424190 with merge 2b8a374...

bors commented:
:broken_heart: Test failed - [auto-win-gnu-32-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/2377)

petrochenkov commented:
@bors retry

bors commented:
:hourglass: Testing commit c424190 with merge 33b1788...

bors commented:
:broken_heart: Test failed - [auto-win-gnu-32-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/2382)

alexcrichton commented:
@bors: retry

On Fri, Sep 2, 2016 at 8:38 AM, bors <notifications@github.com> wrote:

> 💔 Test failed - auto-win-gnu-32-opt-rustbuild
> <https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/2382>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/36016#issuecomment-244410430>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAD95LEttJbZ3kj8yHLKQMY2cMhKgN_Wks5qmEMSgaJpZM4JuP6x>
> .
>

bors commented:
:hourglass: Testing commit c424190 with merge 5004088...

bors commented:
:broken_heart: Test failed - [auto-win-msvc-64-cargotest](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/1711)

alexcrichton commented:
@bors: retry

On Fri, Sep 2, 2016 at 9:08 AM, bors <notifications@github.com> wrote:

> 💔 Test failed - auto-win-msvc-64-cargotest
> <https://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/1711>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/36016#issuecomment-244418265>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAD95IVh14j9ihNjaEwC163fgfc2oBcdks5qmEoTgaJpZM4JuP6x>
> .
>

joshtriplett commented:
The builder seems to be failing on various locked git repositories (unrelated to this change).  Any idea what's going on with the builder?

bors commented:
:lock: Merge conflict

bors commented:
:umbrella: The latest upstream changes (presumably #35957) made this pull request unmergeable. Please resolve the merge conflicts.

petrochenkov commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit 436cfe5 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 436cfe56534b405786816d4bbcccd11ed7571981 -->

bors commented:
:hourglass: Testing commit 436cfe5 with merge d748fa6...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/10414), [auto-linux-32-opt](https://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/10468), [auto-linux-64-cargotest](https://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/1702), [auto-linux-64-cross-freebsd](https://buildbot.rust-lang.org/builders/auto-linux-64-cross-freebsd/builds/1527), [auto-linux-64-debug-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/3686), [auto-linux-64-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/10408), [auto-linux-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/10369), [auto-linux-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/2341), [auto-linux-64-x-android-t](https://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/10346), [auto-linux-cross-opt](https://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/3568), [auto-linux-musl-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/5518), [auto-mac-32-opt](https://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/10447), [auto-mac-64-opt](https://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/10420), [auto-mac-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/2365), [auto-mac-cross-ios-opt](https://buildbot.rust-lang.org/builders/auto-mac-cross-ios-opt/builds/1544), [auto-win-gnu-32-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/5534), [auto-win-gnu-32-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/2396), [auto-win-gnu-64-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/5510), [auto-win-msvc-32-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/4703), [auto-win-msvc-64-cargotest](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/1721), [auto-win-msvc-64-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/5515), [auto-win-msvc-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/2391)
Approved by: nikomatsakis
Pushing d748fa6ecccf6f5b4c7ae4abee0a2d206b165de2 to master...

joshtriplett commented:
Awesome work, @petrochenkov!
